ext/page_log/palm/palm.c

```diff
@@ -354,6 +365,71 @@ palm_kv_err(PALM *palm, WT_SESSION *session, int ret, const char *format, ...)
+  WT_PAGE_LOG_ENCRYPTION *encrypt_out)
+{
+    static WT_PAGE_LOG_ENCRYPTION zero_encryption;
+    WT_PAGE_LOG_ENCRYPTION tmp;
+    bool was_zeroed;
+
+    /*
+     * The DEK is an encrypted encryption key. A production implementation of the page log interface
+     * would do encryption, using the DEK when it is set. If the DEK is not set, the implementation
+     * must figure out what the DEK should be, which may take some time. The DEK is stored with the
+     * page, and when the implementation gets a page it knows how to decrypt it. It also passes the
+     * DEK to the user of the interface (WiredTiger). That DEK must be kept and used for subsequent
+     * deltas to the page. Thus when deltas are written, the DEK doesn't have to be recomputed.
+     *
+     * Here in PALM, we don't want to do any encryption. Since the encrypt/decryption would
+     * invisible to the calling layer (WiredTiger), having encryption doesn't help test WiredTiger
+     * at all. Also, it gets in the way of efficient debugging.
+     *
+     * However, we do want to test that WiredTiger is passing along the DEK whenever it can and
+     * should. If it stopped doing so, the production page log would need to determine the DEK for
+     * itself more often, and we might not notice the error.
+     *
+     * So WiredTiger receives a DEK with every page get. When writing a delta for such a page, it
+     * needs to pass that DEK. One the other hand, when writing a delta for page that WiredTiger
+     * generated and wrote during the current connection, it uses a zeroed DEK, that's the best it
+     * can do.
+     *
+     * To test this without doing any extra KV requests, we generate and store a DEK for any page
+     * write that doesn't already have it - a simple encoding of the table id and page id. Then,
+     * we'd expect that if a DEK is ever passed to us in the put path, it must match that simple
+     * encoding. That tests that the correct DEK is being passed.
+     *
+     * To test that we're passing a DEK when we should, we compare the base_lsn to the LSN we
+     * started the run with. If the base_lsn is less than that, then WiredTiger must have previously
+     * gotten the page from the page log interface, hence the DEK should be set.
+     */
+#define PALM_DEK_FORMAT ("%" PRIu64 ":%" PRIu64)
+    tmp = zero_encryption;
+    if ((size_t)snprintf(&tmp.dek[0], sizeof(tmp.dek), PALM_DEK_FORMAT, table_id, page_id) >
+      sizeof(tmp.dek))
+        assert(false); /* should never overflow */
+
+    was_zeroed = PALM_ENCRYPTION_EQUAL(*encrypt_in, zero_encryption);
+    if (was_zeroed)
+        *encrypt_out = tmp;
+    else {
+        if (!PALM_ENCRYPTION_EQUAL(*encrypt_in, tmp))
+            return (palm_err(palm, session, EINVAL,
+              "encryption dek %31s does not match expected value %31s", encrypt_in->dek, tmp.dek));
+        PALM_VERBOSE_PRINT(palm, "palm using saved dek: %s\n", encrypt_in->dek);
```

Comment:
I think treating this as a string is generally not the right approach. The `%31s` format string hides the case where the DEK has a mismatch in the final element, and wouldn't it also terminate early if any of the members are zero?

Reply:
When we do anything with dek "for real", we always use structure copy or memcmp, never string operations.  The printing of dek as strings is only in error messages or verbose prints - which are not strictly necessary, but may be helpful for debugging.  When PALM is in use, then PALM is generating the deks (for testing), and these are all small strings, like "200:50".  I don't think we need to get any more meticulous with printing these...?

Reply:
Yeah, I'm comfortable with all the real operations. I'm a little suspicious of the print just because a comparison would fail (if the last byte got clobbered somewhere, for example) but the print wouldn't show why. I agree for the problem domain you've stated, it's not something to deal with right now.